File: app/src/main/java/com/handstandsam/maintainableespresso/features/itemdetail/ItemDetailPresenter.java
package com.handstandsam.maintainableespresso.features.itemdetail;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;

import com.handstandsam.maintainableespresso.di.NetworkModule;
import com.handstandsam.maintainableespresso.models.Item;
import com.handstandsam.maintainableespresso.repository.CheckoutCart;
import com.handstandsam.maintainableespresso.repository.SessionManager;

import javax.inject.Inject;

public class ItemDetailPresenter {

    public static final String BUNDLE_PARAM_ITEM = "item";
    private final ItemDetailActivity.ItemDetailView view;

    @Inject
    SessionManager sessionManager;

    @Inject
    CheckoutCart cart;

    Item item;

    public ItemDetailPresenter(@NonNull ItemDetailActivity.ItemDetailView view) {
        this.view = view;
        view.getAppComponent().inject(this);
    }

    public void onResume(Intent intent) {

        Bundle extras = intent.getExtras();
        Item item = (Item) extras.get(BUNDLE_PARAM_ITEM);
        this.item = item;

        view.setLabel(item.getLabel());
        view.setImageUrl(item.getImage());
        view.setActionBarTitle(item.getLabel());
    }

    public void addToCardClicked() {
        cart.addItem(item);
        view.showToast(item.getLabel() + " added to cart.");
    }
}

File: app/src/main/java/com/handstandsam/shoppingapp/features/itemdetail/ItemDetailPresenter.kt
package com.handstandsam.shoppingapp.features.itemdetail

import android.content.Intent
import com.handstandsam.shoppingapp.cart.ShoppingCart
import com.handstandsam.shoppingapp.models.Item
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.io.Serializable

class ItemDetailPresenter(
    private val view: ItemDetailActivity.ItemDetailView,
    private var cart: ShoppingCart
) {
    internal lateinit var item: Item

    fun onResume(intent: Intent) {
        val extras = intent.extras
        val item = extras!!.get(BUNDLE_PARAM_ITEM) as ItemDetailData
        this.item = item.toItem()

        view.setLabel(item.label)
        view.setImageUrl(item.image)
        view.setActionBarTitle(item.label)
    }

    fun addToCardClicked() {
        CoroutineScope(Dispatchers.IO).launch {
            cart.incrementItemInCart(item)
        }
    }

    companion object {
        const val BUNDLE_PARAM_ITEM = "item"
    }
}
